src/backend/Level0/CHIPBackendLevel0.cc

```diff
@@ -785,139 +820,98 @@ void CHIPQueueLevel0::initializeQueueGroupProperties() {
-                                 CHIPQueueFlags Flags, int Priority)
-    : CHIPQueue(ChipDev, Flags, Priority) {
-  ze_result_t Status;
-  auto ChipDevLz = ChipDev;
-  auto Ctx = ChipDevLz->getContext();
-  auto ChipContextLz = (CHIPContextLevel0 *)Ctx;
-
-  ZeCtx_ = ChipContextLz->get();
-  ZeDev_ = ChipDevLz->get();
-
-  logTrace("CHIPQueueLevel0 constructor called via Flags and Priority");
-  initializeQueueGroupProperties();
-
-  ze_command_queue_desc_t QueueDescriptor = getNextComputeQueueDesc();
-
-  switch (Priority_) {
+ze_command_queue_desc_t CHIPDeviceLevel0::getQueueDesc_(int Priority) {
+  ze_command_queue_desc_t QueueDesc = {ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC,
+                                       nullptr, // pNext
+                                       0,       // ordinal
+                                       0,       // index
+                                       0,       // flags
+                                       ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS,
+                                       ZE_COMMAND_QUEUE_PRIORITY_NORMAL};
+
+  switch (Priority) {
   case 0:
-    QueueDescriptor.priority = ZE_COMMAND_QUEUE_PRIORITY_PRIORITY_HIGH;
+    QueueDesc.priority = ZE_COMMAND_QUEUE_PRIORITY_PRIORITY_HIGH;
     break;
   case 1:
-    QueueDescriptor.priority = ZE_COMMAND_QUEUE_PRIORITY_NORMAL;
+    QueueDesc.priority = ZE_COMMAND_QUEUE_PRIORITY_NORMAL;
     break;
   case 2:
-    QueueDescriptor.priority = ZE_COMMAND_QUEUE_PRIORITY_PRIORITY_LOW;
+    QueueDesc.priority = ZE_COMMAND_QUEUE_PRIORITY_PRIORITY_LOW;
     break;
   default:
     CHIPERR_LOG_AND_THROW(
         "Invalid Priority range requested during L0 Queue init", hipErrorTbd);
   }
 
-  std::lock_guard<std::mutex> LockQueues(Backend->QueueCreateDestroyMtx);
-  Status = zeCommandQueueCreate(ZeCtx_, ZeDev_, &QueueDescriptor, &ZeCmdQ_);
-  CHIPERR_CHECK_LOG_AND_THROW(Status, ZE_RESULT_SUCCESS,
-                              hipErrorInitializationError);
-
-#ifdef L0_IMM_QUEUES
-  initializeComputeListImm();
-  initializeCopyListImm();
-#endif
+  return QueueDesc;
 }
 
-CHIPQueueLevel0::CHIPQueueLevel0(CHIPDeviceLevel0 *ChipDev,
-                                 ze_command_queue_handle_t ZeCmdQ)
-    : CHIPQueue(ChipDev, 0, L0_DEFAULT_QUEUE_PRIORITY) {
-  auto ChipDevLz = ChipDev;
-  auto Ctx = ChipDevLz->getContext();
-  auto ChipContextLz = (CHIPContextLevel0 *)Ctx;
+ze_command_queue_desc_t
+CHIPDeviceLevel0::getNextComputeQueueDesc(int Priority) {
 
-  ZeCtx_ = ChipContextLz->get();
-  ZeDev_ = ChipDevLz->get();
+  assert(ComputeQueueGroupOrdinal_ > -1);
+  ze_command_queue_desc_t CommandQueueComputeDesc = getQueueDesc_(Priority);
+  CommandQueueComputeDesc.ordinal = ComputeQueueGroupOrdinal_;
 
-  initializeQueueGroupProperties();
+  auto MaxQueues = ComputeQueueProperties_.numQueues;
+  std::lock_guard<std::mutex> LockQueueIndexIterator(NextQueueIndexMtx_);
+  NextComputeQueueIndex_ = (NextComputeQueueIndex_ + 1) % MaxQueues;
```

Comment:
Missed index update CommandQueueComputeDesc.index = NextComputeQueueIndex_; and then increment. Both should be under lock.

Reply:
Index starts at 0 during initialization